mc/query/eval.go

```diff
@@ -0,0 +1,372 @@
+package query
+
+import (
+	"fmt"
+	pb "github.com/mediachain/concat/proto"
+	"strings"
+)
+
+// query evaluation: very primitive eval with only simple statements
+// will have to do until we have an index and we can compile to sql.
+func EvalQuery(query *Query, stmts []*pb.Statement) ([]interface{}, error) {
+	nsfilter := makeNamespaceFilter(query)
+
+	cfilter, err := makeCriteriaFilter(query)
+	if err != nil {
+		return nil, err
+	}
+
+	rs, err := makeResultSet(query)
+	if err != nil {
+		return nil, err
+	}
+
+	rs.begin(len(stmts))
+	for _, stmt := range stmts {
+		if nsfilter(stmt) && cfilter(stmt) {
+			rs.add(stmt)
+		}
+	}
+	rs.end()
+
+	return rs.result(), nil
+}
+
+type QueryResultSet interface {
+	begin(hint int)
+	add(*pb.Statement)
+	end()
+	result() []interface{}
+}
+
+type QueryEvalError string
+
+func (e QueryEvalError) Error() string {
+	return string(e)
+}
+
+type StatementFilter func(*pb.Statement) bool
+type ValueCriteriaFilter func(*pb.Statement, string) bool
+type TimeCriteriaFilter func(*pb.Statement, int64) bool
+type CompoundCriteriaFilter func(stmt *pb.Statement, left, right StatementFilter) bool
+
+func idCriteriaFilter(stmt *pb.Statement, id string) bool {
+	return stmt.Id == id
+}
+
+func publisherCriteriaFilter(stmt *pb.Statement, pub string) bool {
+	return stmt.Publisher == pub
+}
+
+func sourceCriteriaFilter(stmt *pb.Statement, src string) bool {
+	// only support simple statements for now, so src = publisher
+	return stmt.Publisher == src
+}
+
+var valueCriteriaFilters = map[string]ValueCriteriaFilter{
+	"id":        idCriteriaFilter,
+	"publisher": publisherCriteriaFilter,
+	"source":    sourceCriteriaFilter}
+
+func timestampFilterLTEQ(stmt *pb.Statement, ts int64) bool {
+	return stmt.Timestamp <= ts
```

Comment:
They are kind of primops, and thus have an implementation; it's also the most straightforward way to do it.